### PR TITLE
fix: fp8 quantization failure of qwen 2.5 VL 7B model

### PR DESCRIPTION
**Pull request by PanJason**
## Motivation

Fix this bug https://github.com/sgl-project/sglang/issues/6828 of qwen 2.5 VL 7B where fp8 quantization was reported to fail in the visual model part.

## Modifications

1. Make the initialization of the intermediate size in qwen 2.5 VL 7B 8-aligned
2. Change both `weight_loader` and `weight_loader_v2` in `linear.py` to add zero padding when loading weights that are not 8-aligned. After this fix, both `w8a8_fp8` and `fp8` can work
3. Change `w8a8_int8.py` to fix the runtime crash when using `w8a8_int8`

We don't change the kernel as the previous version did #7448

## Remaining work
While fixing the quantization failure for qwen2_5_VL model, I found two other issues. Since they are not exactly related to VL model, I created two other issues. The fix of them will go to different patch series
1. When sending two concurrent requests to qwen2_5_VL model, rotary_embedding crashed regardless of quantization, this is reported in #10623 
2. When fixing w8a8_int8, I found qwen2.5 7B instruct base model generates gibberish. This is reported in #10626 

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.ai/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.ai/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.ai/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.ai/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.ai/developer_guide/contribution_guide.html#benchmark-the-speed).

### CODE REVIEW

File: python/sglang/srt/layers/linear.py [hunk 1]
```diff
@@ -626,9 +626,38 @@ def weight_loader(
...
+                            )
+                        else:
+                            # All padding
+                            pad_shape = list(loaded_weight.shape)
+                            pad_shape[output_dim] = shard_size
+                            loaded_weight = torch.zeros(
+                                pad_shape,
+                                dtype=loaded_weight.dtype,
+                                device=loaded_weight.device,
+                            )
```
Comment (JustinTong0323) [inline python/sglang/srt/layers/linear.py]: [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself). Refactor it into a helper function.

**Review by JustinTong0323 — CHANGES_REQUESTED** [review]

**Comment by PanJason** [comment]
Refactor as requested. Now fix fp8 and w8a8_int8

**Comment by PanJason** [comment]
@JustinTong0323 @mickqian Could you please check this PR?

**Comment by b8zhong** [comment]
Can you check if the MMMU accuracy is recovered?

**Comment by Swipe4057** [comment]
Hi, I tested it on Qwen2.5-7B-VL in w8a8_fp8 quantization and it works without errors, checked on local corporate benchmark - the accuracy dropped by about 3 %.

**Comment by PanJason** [comment]
> Can you check if the MMMU accuracy is recovered?

Can you point me to where the MMMU accuracy tests are? Are they local or run in CI?

**Comment by PanJason** [comment]
> Hi, I tested it on Qwen2.5-7B-VL in w8a8_fp8 quantization and it works without errors, checked on local corporate benchmark - the accuracy dropped by about 3 %.

Is 3% comparing before and after quantization?

**Comment by Swipe4057** [comment]
> > Hi, I tested it on Qwen2.5-7B-VL in w8a8_fp8 quantization and it works without errors, checked on local corporate benchmark - the accuracy dropped by about 3 %.
> 
> Is 3% comparing before and after quantization?

yes

**Comment by b8zhong** [comment]
> Can you point me to where the MMMU accuracy tests are? Are they local or run in CI?

https://github.com/sgl-project/sglang/blob/main/benchmark/mmmu/README.md

**Comment by PanJason** [comment]
@b8zhong Hi I just tested the benchmark on my local machine with no quantization, `w8a8_fp8` quantization and `fp8` quantization. The following is the command I used for server. I dropped `--enable-mixed-chunk` because of #10623 .
```bash
python3 -m sglang.launch_server --model-path Qwen/Qwen2.5-VL-7B-Instruct  --tp 1 --dp 1 --port 8001 --host 0.0.0.0 \ 
--mem-fraction-static 0.9 --schedule-conservativeness 0.3 --grammar-backend xgrammar --enable-metrics \ 
--allow-auto-truncate --enable-multimodal --mm-attention-backend fa3 --attention-backend flashinfer \ 
--schedule-policy fcfs
```

On the client side, the same concurrency level as https://github.com/sgl-project/sglang/blob/main/benchmark/mmmu/README.md
```bash
python3 bench_sglang.py --port 8001 --concurrency 16
```
## Results:
| Quantization    | Overall accuracy |
| -------- | ------- |
| No quantization  | 0.516    |
| w8a8_fp8 |  0.506     |
| fp8    |  0.507    |

## Test env
[test.txt](https://github.com/user-attachments/files/22460767/test.txt)

Let me know if anything else needs to be done for this patch

**Review by mickqian — APPROVED** [review]

**Review by JustinTong0323 — APPROVED** [review]
LGTM.

**Review by BBuf — APPROVED** [review]
